part-persist: implement message aggregation

AxelSchneewind opened the pull request:
Created a component for partitioned communication that supports message aggregation, based on the existing part-persist component. The goal is to prevent drops in effective bandwidth when using too fine-grained partitionings.

This component allows enforcing hard limits on size and count of transferred partitions, regardless of the partitioning required by the application. These limits can be specified using mca-parameters (min_message_size, max_message_count). Their default values might require revision.

If a user-provided partitioning violates the constraints, a more coarse-grained partitioning is selected, where multiple user-partitions are mapped to an internal (transfer) partition. Transfer of an internal partition is started as soon as Pready has been called on all corresponding user-partitions. Each transfer partition is associated with an atomic counter, tracking the number of corresponding user-partitions that have been marked ready.

The implementation could also be extended to use more advanced aggregation algorithms.

This implementation is a result of "Benchmarking the State of MPI Partitioned Communication in Open MPI" presented at EuroMPI 2024 (https://events.vsc.ac.at/event/123/page/341-program).

mdosanjh commented:
I’ll look into it today.

On Tue, Jan 28, 2025 at 09:15 Tommy Janjusic ***@***.***>
wrote:

> @mdosanjh <https://github.com/mdosanjh> Any chance we could get an
> initial review from you on this PR?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/13039#issuecomment-2619454316>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4RGBD4MSPZA3QM3Y4HJZ32M6UJNAVCNFSM6AAAAABVHNG5I6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJZGQ2TIMZRGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hppritcha commented:
i think someone needs to look at the mpi4py failure.  looks specific to changes in this PR.

cniethammer commented:
Just an assumption from my side having a look at the error message without diving into the mpi4py code: It could be related to the extension of fields in the Request handle, which is mapped to a python object in mpi4py.

hppritcha commented:
I doubt that.  Probably failing only with mpi4py because it has the most extensive test suite in our CI infrastructure.  This pr should not be merged till mpi4py passes.

AxelSchneewind commented:
Thank you very much, I will look into it. That error never occurred with my tests but maybe I missed an edge case

AxelSchneewind commented:
The bug that was found by the mpi4py-tests pipeline should now be resolved.
The error was caused by a mismatch between function pointers for the implementations of `MPI_Start` and `MPI_P[send|recv]_Init` (for more detail, see below).

I solved it by using separate request lists for each component.

This might also be relevant for PR #12796 that also introduces a new part-component.
I understand the issue as follows:
- The function pointer for `MPI_Psend_Init` is stored in the `mca_part`-object and set by `mca_part_base_select`.
- The function pointer for `MPI_Start` is stored in the request objects (`req_start`-field), which are preallocated when opening a component. As I understand it, all components  (i.e. both `part_persist` and `part_persist_aggregated`) are opened during `MPI_Init`, and initialize some request objects each.

The segfault can occur when a request object is used where `req_start` points to the implementation from a component that has not been selected

AxelSchneewind commented:
bot:retest

hppritcha commented:
@mdosanjh could you review this PR?

mdosanjh commented on ompi/mca/part/persist_aggregated/part_persist_aggregated.h:
```diff
@@ -0,0 +1,717 @@
+        *internal_partitions = partitions / *factor;
+        *remainder = partitions % (*internal_partitions);
+
+        if (*remainder == 0) { // we still need the size of last partition
```
Coding conventions would have this as 0 == *remainder.

mdosanjh reviewed (changes requested):
There are a couple big things to sort out here: one is to squash this into a single commit. The second is that I'm wondering if this should be integrated or essentially replace the existing persist module. I'm not seeing a down side to adding aggregation  and I think it would be cleaner to have the one module built on top of persistent communication. One minor thing I've found so far is mentioned in the comment below.

AxelSchneewind committed:
implement message aggregation for part/persist

This aggregation scheme is intended to allow OpenMPI to transfer larger messages
if the user-reported partitions are too small or too many.
This is achieved by using an internal partitioning where each internal (transfer)
partition corresponds to one or multiple user-reported partitions.

The implementation provides an interface for insertion of user partitions,
that optionally outputs a transfer partition that is ready.

Internally, each transfer partition is associated with an atomic counter,
tracking the number of corresponding pready-calls. As soon as a counter reaches
the number of corresponding user-partitions, the transfer partition is returned
in the respective insertion call.

Signed-off-by: Axel Schneewind <axel.schneewind@hlrs.de>

AxelSchneewind commented:
@mdosanjh thanks for the feedback. I'll squash it into one commit.

About the second point: I added it as a new component because I wasn't sure if it's better to have it as an opt-in initially.
But I agree, having one component is much cleaner and I also don't see any problems integrating it into part-persist directly. 
I will do that.

AxelSchneewind commented:
Integrated the changes into part-persist and squashed it into one commit.
Also, removed some small unnecessary changes.

hppritcha commented:
@mdosanjh could you rewiew again?

hppritcha commented:
Anyone working on this PR?  Should we close it?  It needs to rebased at least to resolve conflicts.

cniethammer commented:
Thanks for having a look. We'll rebase this on the latest main and ping asking for another review when done.

AxelSchneewind commented:
Rebased onto main. The PR is ready for review again

hppritcha commented:
@AxelSchneewind looks like you need to rebase again.

mdosanjh commented on ompi/mca/part/persist/part_persist.h:
```diff
@@ -78,13 +80,72 @@ struct ompi_part_persist_t {
+
+    if (part_size < min_part_size) {
+        // solve partitions = (internal_partitions - 1) * factor + remainder for factor and remainder
+        _factor = min_part_size / part_size;
```
I would like some comments here explaining the logic; I'm following it through and it generally makes sense (though if I'm reading everything correctly, min_part_size is actually 2x min part size? (in a min part size of 199 and part_size of 100, the factor still becomes 1)

AxelSchneewind replied:
Youre right, I should probably change that. Using ceiled division to compute the factor should be sufficient.

I'll also add more explanations here.
